cumulusci/tasks/bulkdata/utils.py

```diff
@@ -66,82 +47,13 @@ def setup_epoch(inspector, table, column_info):
-            return "Aborted", None
-        elif "InProgress" in statuses or "Queued" in statuses:
-            return "InProgress", None
-        elif "Failed" in statuses:
-            return "Failed", state_messages
-
-        failures = tree.find(f".//{{{ns}}}numberRecordsFailed")
-        if failures is not None:
-            num_failures = int(failures.text)
-            if num_failures:
-                return "CompletedWithFailures", [f"Failures detected: {num_failures}"]
-
-        return "Completed", None
-
-    def _wait_for_job(self, job_id, error_behaviour: str = "raise") -> str:
-        """With for job_id to finish.
-
-        If there are any row errors, error_behaviour says what happens.
-
-        If its "raise" (the default), then throw an exception.
-
-        If its "return" then return "fail" and set an instance variable: self.error_messages
-        """
-        assert error_behaviour in ("raise", "return")
-        while True:
-            job_status = self.bulk.job_status(job_id)
-            self.logger.info(
-                f"    Waiting for job {job_id} ({job_status['numberBatchesCompleted']}/{job_status['numberBatchesTotal']})"
-            )
-            result, messages = self._job_state_from_batches(job_id)
-            if result != "InProgress":
-                break
-            time.sleep(10)
-        self.logger.info(f"Job {job_id} finished with result: {result}")
-        if "Fail" in result:
-            for state_message in messages:
-                self.logger.error(f"Batch failure message: {state_message}")
-            if error_behaviour == "raise":
-                raise BulkDataException("Job Error", messages)
-            else:
-                self.error_messages = messages
-
-        return result
-
-    def _sql_bulk_insert_from_csv(self, conn, table, columns, data_file):
-        if conn.dialect.name in ("postgresql", "psycopg2"):
-            # psycopg2 (the postgres driver) supports COPY FROM
-            # to efficiently bulk insert rows in CSV format
-            with conn.connection.cursor() as cursor:
-                cursor.copy_expert(
-                    f"COPY {table} ({','.join(columns)}) FROM STDIN WITH (FORMAT CSV)",
-                    data_file,
-                )
-        else:
-            # For other db drivers we need to use standard SQL
-            # -- this is optimized for ease of implementation
-            # rather than performance and may need more work.
-            reader = unicodecsv.DictReader(data_file, columns)
-            table = self.metadata.tables[table]
-            rows = list(reader)
-            if rows:
-                conn.execute(table.insert().values(rows))
+class SqlAlchemyMixin:
+    def _sql_bulk_insert_from_records(self, conn, table, columns, records):
+        table = self.metadata.tables[table]
+
+        for batch in BatchIterator(records, n=100):
+            conn.execute(table.insert(), [dict(zip(columns, row)) for row in batch])
```

Comment:
This was Jason's solution to get EDAEASY to work at all (I had a similar problem with NPSP's new data set). 

It seems reasonable to make it conditional on the underlying database type, but I don't know the field of options well at all. Are you aware of any other backends that require the same workaround?

Reply:
Not sure about other databases. For SQLite I think it actually depends on a compile-time setting :-/ https://stackoverflow.com/a/15313495

Reply:
Why is it 100 in particular? The links all say 999 is the max?

In general there is something deeply non-intuitive about all of this string parsing and generating in a single process, but that's mostly independent of David's refactoring. It's just literally the way that SQL Alchemy and SQLite work. It's almost as weird in Snowfakery. Generating strings for talking to an in-process library is kind of weird. Being limited to 999 inserts at the same time is even weirder.

Reply:
I think this should be refined, but I'm not 100% sure how to trigger the error in the first place, or how to reason about it in the code. We set the limit at 100 _records_, arbitrarily, because that's what Jason found by experiment worked on EDAEASY. The SQLite limit I believe is on variables in the INSERT statement.

Reply:
Because I could not reproduce the original error with or without it, I removed this workaround for the moment. Once I can document how to reproduce it, I'd like to do a new revision with a chunking strategy that's less ad-hoc.